src/OrleansRuntime/GrainDirectory/GrainDirectoryPartition.cs

```diff
@@ -364,21 +369,23 @@ internal void RemoveGrain(GrainId grain)
         internal AddressesAndTag LookUpActivations(GrainId grain)
         {
             var result = new AddressesAndTag();
+            IGrainInfo graininfo;
             lock (lockable)
             {
-                IGrainInfo graininfo;
-                if (partitionData.TryGetValue(grain, out graininfo))
+                if (!partitionData.TryGetValue(grain, out graininfo))
                 {
-                    result.Addresses = new List<ActivationAddress>();
-                    result.VersionTag = partitionData[grain].VersionTag;
-
-                    foreach (var route in partitionData[grain].Instances)
-                    {
-                        if (IsValidSilo(route.Value.SiloAddress))
-                            result.Addresses.Add(ActivationAddress.GetAddress(route.Value.SiloAddress, grain, route.Key));
-                    }
+                    return result;
                 }
             }
+
+            result.Addresses = new List<ActivationAddress>();
```

Comment:
Not sure how much this will help, but we could pass `graininfo.Instances.Count` to the constructor here to set the capacity and avoid additional allocations/copies.